dmriprep/utils/vectors.py

```diff
@@ -0,0 +1,334 @@
+from __future__ import division, print_function
+import os
+import nibabel as nib
+import numpy as np
+
+
+def make_gtab(fbval, fbvec, vector_dir, final, b0_thr=50):
+    """
+    Create gradient table from bval/bvec.
+
+    Parameters
+    ----------
+    fbval : str
+        File path of the b-values.
+    fbvec : str
+        File path of the b-vectors.
+    vector_dir : str
+        Path to vectors directory in subject output.
+    final : bool
+        Indicates whether to flag the output gradient
+    b0_thr : float
+        Gradient threshold below which volumes and vectors are considered B0's.
+
+    Returns
+    -------
+    gtab_file : str
+        File path to pickled DiPy gradient table object.
+    gtab : Obj
+        DiPy object storing diffusion gradient information.
+    final_fbval : str
+        File path of saved final b-values if `final` is True.
+    final_fbvec : str
+        File path of saved final b-vectors if `final` is True.
+    """
+    import os
+    from dipy.io import save_pickle, read_bvals_bvecs
+    from dipy.core.gradients import gradient_table
+
+    # Read b-values and b-vectors from disk
+    # Files can be either ‘.bvals’/’.bvecs’ or ‘.txt’ or ‘.npy’ (containing
+    # arrays stored with the appropriate values).
+    if fbval and fbvec:
+        bvals, bvecs = read_bvals_bvecs(fbval, fbvec)
+    else:
+        raise OSError("Either bvals or bvecs files not found (or rescaling failed)!")
+
+    # Creating the gradient table
+    gtab = gradient_table(bvals, bvecs)
+
+    # Correct b0 threshold
+    gtab.b0_threshold = b0_thr
+
+    # Correct b0 mask
+    gtab_bvals = gtab.bvals.copy()
+    b0_thr_ixs = np.where(gtab_bvals < gtab.b0_threshold)[0]
+    gtab_bvals[b0_thr_ixs] = 0
+    gtab.b0s_mask = gtab_bvals == 0
+
+    # Show info
+    print(gtab.info)
+
+    # Save bval/bvec text file equivalents in the case that these are the final
+    # versions of our gradient vectors
+    if final is True:
+        final_bval_path = "%s%s" % (vector_dir, '/final_bval.bval')
```

Comment:
Since the function `make_gtab` gets reused each time the vectors are updated throughout the pipeline (this can happen at least 2-3 times), we can save the result with sequential integers as file suffices (this allows us to keep track of each change made to the vectors and in what order), but we can also use a `final` trigger that allows us to distinguish for the user the final version of the vectors to be used for post-preprocessing analysis. So by final here I am indeed referring to late stage of analysis.

Reply:
I would allow nipype to keep track of this.